### PR TITLE
D3ASIM-2925: Added energy forcast endpoint for external pv and load.

### PR DESCRIPTION
**Pull request by hannesdiedrich**

### CODE REVIEW

File: src/d3a/models/strategy/external_strategies/load.py
Comment (faizan2590) [inline src/d3a/models/strategy/external_strategies/load.py:49]: Any specific reason to wrap same private property to public?

File: src/d3a/models/strategy/external_strategies/load.py
Comment (faizan2590) [inline src/d3a/models/strategy/external_strategies/load.py:367]: Since all strategies have unique channel prefix, we can  generalise the suffix like here `set_load_power_forecast` to simply `power_forecast`

File: src/d3a/models/strategy/external_strategies/pv.py
Comment (faizan2590) [inline src/d3a/models/strategy/external_strategies/pv.py:242]: Good to be abstracted into a method but can you help me understand the need to be called from `event_activate ` too?

File: src/d3a/models/strategy/external_strategies/load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:367]: Agreed!

File: src/d3a/models/strategy/external_strategies/load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:49]: Yes, in order to overwrite the `self._channel_dict` in the LoadForecastExternalStrategy.

File: src/d3a/models/strategy/external_strategies/pv.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/pv.py:242]: I havent added it to the `event_activate`. Or can you point me to it please?

File: src/d3a/models/strategy/external_strategies/load.py
Comment (faizan2590) [reply src/d3a/models/strategy/external_strategies/load.py:49]: Why not re-signature the existing `_channel_dict ` to `channel_dict ` ?

File: src/d3a/models/strategy/predefined_load.py
Comment (faizan2590) [inline src/d3a/models/strategy/predefined_load.py:138]:
```suggestion
    parameters = ('power_forecast_W', 'fit_to_limit', 'energy_rate_increase_per_update',
```

File: src/d3a/models/strategy/predefined_load.py
Comment (faizan2590) [inline src/d3a/models/strategy/predefined_load.py:180]:
```suggestion
        assert energy_forecast_Wh >= 0.0, energy_forecast_Wh cannot be a negative value.
```
Minor cosmetic.

File: src/d3a/models/strategy/external_strategies/load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:49]: Because i need both. Please have a look at the LoadForecastExternalStrategy.
 I can not update a property (`channel_dict`), but i can change the 
`_channel_dict` which is not a property.

File: src/d3a/models/strategy/predefined_pv.py
Comment (faizan2590) [inline src/d3a/models/strategy/predefined_pv.py:181]:
```suggestion
    parameters = ('power_forecast_W', 'panel_count', 'initial_selling_rate', 'final_selling_rate',
```

File: src/d3a/models/strategy/predefined_load.py
Comment (faizan2590) [reply src/d3a/models/strategy/predefined_load.py:180]: On second thought, why not raise an exception, so that simulation doesn't crash due to external clients' wrong value ?
```suggestion
        if energy_forecast_Wh <= 0.0:
            raise MarketException("energy_forecast_Wh cannot be a negative value.")
```

File: src/d3a/models/strategy/predefined_pv.py
Comment (faizan2590) [inline src/d3a/models/strategy/predefined_pv.py:214]: [same](https://github.com/gridsingularity/d3a/pull/931/files#r511863864) applies here too.

File: src/d3a/models/strategy/predefined_pv.py
Comment (faizan2590) [inline src/d3a/models/strategy/predefined_pv.py:216]: Minor, to be removed before merging.

File: src/d3a/models/strategy/predefined_load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/predefined_load.py:180]: Its a matter of taste really. 
1. An AssertionError behaves the same as an Exception AFAIK
2. The user can not set a negative value, please have a look into the `_set_power_forecast_impl`, there it is already checked (and does not make the simulation crash because it is caught by the try/except.

TLDR; will raise an Exception as you suggested.

File: src/d3a/models/strategy/external_strategies/load.py
Comment (faizan2590) [reply src/d3a/models/strategy/external_strategies/load.py:49]: umm, both are property.

Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:49]: oh ,yes, sorry, i had so many forth and backs here that i lost track. 
But here too it is a matter of taste. I either do as I did or I have to overwrite the event_activate method in the LoadForecastExternalStrategy. 
Maybe @spyrostz  is best to be ask here for the best practice.

File: src/d3a/models/strategy/external_strategies/load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:367]: @faizan2590 Sorry, I didn't read properly the first time.
I thought about this when implementing too, and then the `_set_power_forecast` could even live in the ExternalMixin.
However, in the `_set_power_forecast_impl` different Strategy attributes are set and this shows also that pv and load are not identical here. This is why I kept the difference between load and pv.
But if @spyrostz agrees, I will refactor it as you said.

Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:367]: @faizan2590 Actually, no, i think yours is a goof idea and will try to implement it like this.

File: src/d3a/models/strategy/external_strategies/pv.py
Comment (faizan2590) [reply src/d3a/models/strategy/external_strategies/pv.py:242]: Sorry yes you are right, my bad.

File: src/d3a/models/strategy/external_strategies/load.py
Comment (faizan2590) [reply src/d3a/models/strategy/external_strategies/load.py:367]: > I thought about this when implementing too, and then the _set_power_forecast could even live in the ExternalMixin.

As per MRO `_set_power_forecast ` would be overwritten by `LoadForecastExternalStrategy` and `PVForecastExternalStrategy` respectively, so I don't  think you should be worried about base implementation in `ExternalMixin `

Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:367]: Og course this includes to remove the implementations in `LoadForecastExternalStrategy` and `PVForecastExternalStrategy`, currently trying it.

File: src/d3a/models/strategy/predefined_load.py
Comment (faizan2590) [reply src/d3a/models/strategy/predefined_load.py:180]: You can remove it at all from here as it is already covered in `_set_power_forecast_impl `.

Comment (hannesdiedrich) [reply src/d3a/models/strategy/predefined_load.py:180]: yeah, i wanted to keep it as extra layer of security (in case something goes wrong in the conversion between power and energy) But you are right, its not needed as there is just a factor applied.

File: src/d3a/models/strategy/external_strategies/load.py
Comment (spyrostz) [reply src/d3a/models/strategy/external_strategies/load.py:49]: I would actually remove the private property and keep only the public property, unless I am not seeing something.

File: src/d3a/models/strategy/predefined_load.py
Comment (spyrostz) [inline src/d3a/models/strategy/predefined_load.py:135]: I am fairly certain that we will never use this strategy without an external API. I would recommend to merge it with the LoadForecastExternalStrategy

File: src/d3a/models/strategy/predefined_pv.py
Comment (spyrostz) [inline src/d3a/models/strategy/predefined_pv.py:177]: Same as above, we'll never need to instantiate it without an external connection

File: src/d3a/models/strategy/pv.py
Comment (spyrostz) [inline src/d3a/models/strategy/pv.py:101]: Are you certain ? This will have the effect that the produced_energy_forecast will not be updated if e.g. the panel_count parameter is modified, which does not look right to my eyes

**Review by spyrostz — COMMENTED** [review]
Added some comments, will have another look once dealt with

File: src/d3a/models/strategy/predefined_load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/predefined_load.py:135]: ok, makes sense!

File: src/d3a/models/strategy/pv.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/pv.py:101]: Ah yes, you are right, i didnt consider the panel count as an energy related parameter. But it certainly is.

File: src/d3a/models/strategy/external_strategies/load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:49]: ok guys, i gave it yet another look. I dont want to subscribe to `f'{self.channel_prefix}/set_power_forecast'` channel for the normal `LoadExternalMixin`. So i have to change the list of channels in the `LoadForecastExternalStrategy`. I dont see another way of doing it the way i did. I can not overwrite the `event_activate` because i need to call the event_activate of the LoadHours Strategy there.
But maybe there is another way, could you please tell me? Thanks

Comment (faizan2590) [reply src/d3a/models/strategy/external_strategies/load.py:49]: Updating channel dict inside the constructor of this new class could be an option like `self.channel_dict.update({f'{self.channel_prefix}/set_load_power_forecast': self._set_power_forecast})`

Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:49]: No, this wont work, as it is a property. You can not overwrite a property.

Comment (faizan2590) [reply src/d3a/models/strategy/external_strategies/load.py:49]: This property returns dictionary to which you are updating, I haven't tried yet but that was from the top of my head to give a try

Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:49]: I have tried it in the past. 
yes, indeed, the property returns a dict, but it is really a function. You can not call `.update()` on a function. 
(already pycharm is complaining about it. )

Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:49]: what i could do here is to initiate a dictionary `channel_dict` in the constructor and follow the approach you suggested. (so removing the property function.

Comment (spyrostz) [reply src/d3a/models/strategy/external_strategies/load.py:49]: You can also use `super()`

File: src/d3a/models/strategy/external_strategies/__init__.py
Comment (spyrostz) [inline src/d3a/models/strategy/external_strategies/__init__.py:408]: Might be an overkill but I think that in the long run there needs to be an extra validation here, that will check the power forecast against the nameplate characteristics of the load and PV (essentially the max PV/Load power)

File: src/d3a/models/strategy/external_strategies/load.py
Comment (spyrostz) [inline src/d3a/models/strategy/external_strategies/load.py:413]:
```suggestion
        if req.request_type == "set_power_forecast":
            self._set_power_forecast_impl(req.arguments, req.response_channel)
        else:
            super(). _incoming_commands_callback_selection(req)
```

File: src/d3a/models/strategy/external_strategies/load.py
Comment (spyrostz) [inline src/d3a/models/strategy/external_strategies/load.py:423]: Why do you need to call this again ? This should be called by the _update_energy_requirement (essentially the first time that the energy_requirement is populated), therefore there should be no need to be called from here too.

File: src/d3a/models/strategy/external_strategies/pv.py
Comment (spyrostz) [inline src/d3a/models/strategy/external_strategies/pv.py:414]: Same as with the load, this method can be refactored.

File: src/d3a/models/strategy/external_strategies/pv.py
Comment (spyrostz) [inline src/d3a/models/strategy/external_strategies/pv.py:426]: This calculation exists in many places now. How about refactoring to two utility functions called `convert_energy_to_power` and `convert_power_to_energy` respectively ?

File: src/d3a/models/strategy/load_hours.py
Comment (spyrostz) [inline src/d3a/models/strategy/load_hours.py:324]: Not sure why this was needed too. I would expect that the energy requirement should always be populated with the time_slot value (for the simulation case at least, where this value is populated at the beginning of the simulation)

**Review by spyrostz — COMMENTED** [review]
Some comments, LGTM once dealt with

File: src/d3a/models/strategy/external_strategies/__init__.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/__init__.py:408]: I agree in general, but then we need to add this as an input argument for the new Strategy which will also include a UI change. So i suggest to do it later. Shall I add a ticket?

File: src/d3a/models/strategy/external_strategies/load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:413]: Yes, thanks a lot! I only learned about the value of `super()` last week :-)

File: src/d3a/models/strategy/external_strategies/__init__.py
Comment (spyrostz) [reply src/d3a/models/strategy/external_strategies/__init__.py:408]: Please do so yes, but as I mentioned above this is for the long run, not urgent for the hackathon. Thanks!

File: src/d3a/models/strategy/load_hours.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/load_hours.py:324]: Indeed. Remnant from debugging. Will remove it.

File: src/d3a/models/strategy/external_strategies/load.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/load.py:423]: I was not set, because the `event_activate_energy` is overwritten here. But you are right, it should live in there. Moving it.

**Review by spyrostz — APPROVED** [review]
LGTM

File: src/d3a/models/strategy/external_strategies/__init__.py
Comment (hannesdiedrich) [reply src/d3a/models/strategy/external_strategies/__init__.py:408]: Done https://gridsingularity.atlassian.net/browse/D3ASIM-3010

**Review by faizan2590 — APPROVED** [review]
LGTM

**Pull request merged**
